fpdcc/ccfp-asset-dashboard
# Add programmatic scoring

## Overview

For issues:
- #162 
- #163 

Includes:
- more signals for calculating the scores
- a loading spinner while the assets are saved and the calculations process, because sometimes it takes a while when saving miles and miles of trails
- changed the scoring form to only show the number values of geographic and socio economic scores, so it's not user editable

### Notes
For the socio economic distance scoring, I had trouble aggregating all of the buffered geometries into a single geometry collection (like [we do for the other scoring](https://github.com/fpdcc/ccfp-asset-dashboard/blob/9bafe896e589e0c6d3667e48c740c13a82f1a1b7/asset_dashboard/models.py#L393)). Specifically, I was running into a GIS error where the buffered geoms would have [a ring self-intersection error](https://gis.stackexchange.com/questions/15286/ring-self-intersections-in-postgis). All of the solutions said to set the buffer as zero, which would negate our buffered score.

When debugging, I found that I could find the intersection of each buffered poly type separately, so my code is a workaround where we [loop through all of the geometry types](https://github.com/fpdcc/ccfp-asset-dashboard/blob/96be96d9178efb20a6f04f5dd4b1b84e8b378b04/asset_dashboard/models.py#L381) and find the intersection with the socio economic disinvested areas.

The buffer is 1.11 miles based on [the information found in this gis.stackexchange](https://gis.stackexchange.com/questions/8650/measuring-accuracy-of-latitude-and-longitude/8674#8674).

## Testing Instructions
- visit review app: https://ccfp-asset-d-feature-ge-vbizwj.herokuapp.com
- add some assets to a phase
- save some assets within the cmap areas
  - you can reference this cmap map to get the areas https://www.cmap.illinois.gov/2050/maps/eda
- make sure the geographic zone and socio economic score shows up in the project's overview page
